base/shell/explorer/traydeskbtn.cpp

```diff
@@ -0,0 +1,443 @@
+    m_drawWithDedicatedBackground(FALSE),
+    m_bHovering(FALSE),
+    m_bPressed(FALSE),
+    IsHorizontal(FALSE)
+{
+}
+
+BOOL
+CTrayShowDesktopButton::GetTaskbar(OUT HWND* taskbarWnd)
+{
+    *taskbarWnd = NULL;
+    auto parent = GetParent();
+    if (!parent)
+        return FALSE;
+
+    parent = parent.GetParent();
+    if (!parent)
+        return FALSE;
+
+    HWND wnd = parent.m_hWnd;
+    *taskbarWnd = wnd;
+    return wnd != NULL;
+}
+
+INT CTrayShowDesktopButton::WidthOrHeight() const
+{
+    if (IsThemeActive() && !m_highContrastMode)
+    {
+        if (m_drawWithDedicatedBackground)
+        {
+            if (GetSystemMetrics(SM_TABLETPC))
+            {
+                //TODO: DPI scaling - return logical-to-physical conversion of 24, not fixed value
+                return 24;
+            }
+            else
+                return 15;
+        }
+        else
+        {
+            return max(16 + (IsHorizontal
+                ? (m_ContentMargins.cxLeftWidth + m_ContentMargins.cxRightWidth)
+                : (m_ContentMargins.cyTopHeight + m_ContentMargins.cyBottomHeight)
+            ), 18) + 6;
+        }
+    }
+    else
+    {
+        return max(
+            2 * GetSystemMetrics(SM_CXBORDER) + GetSystemMetrics(SM_CXSMICON),
+            2 * GetSystemMetrics(SM_CYBORDER) + GetSystemMetrics(SM_CYSMICON)
+        );
+    }
+}
+
+HRESULT CTrayShowDesktopButton::DoCreate(HWND hwndParent)
+{
+    DWORD style = WS_CHILD | WS_VISIBLE | WS_CLIPSIBLINGS | BS_DEFPUSHBUTTON;
+    Create(hwndParent, NULL, NULL, style);
+
+    if (!m_hWnd)
+        return E_FAIL;
+
+    bool bIconRetrievalFailed = ExtractIconExW(L"imageres.dll", -IDI_IMAGERES_DESKTOP, NULL, &m_icon, 1) == UINT_MAX;
+    if (bIconRetrievalFailed || !m_icon)
+        ExtractIconExW(L"shell32.dll", -IDI_SHELL32_DESKTOP, NULL, &m_icon, 1);
+    
+    m_szIcon.cx = 16;
+    m_szIcon.cy = 16;
+    
+    if (m_icon)
+    {
+        ICONINFO iconInfo;
+        if (GetIconInfo(m_icon, &iconInfo))
+        {
+            BITMAP bmpIcon;
+            if (GetObjectW(iconInfo.hbmMask, sizeof(bmpIcon), &bmpIcon))
+            {
+                m_szIcon.cx = bmpIcon.bmWidth;
+                m_szIcon.cy = bmpIcon.bmHeight;
+                /*
+                If this structure defines a black and white icon, this bitmask is formatted
+                so that the upper half is the icon AND bitmask and the lower half is
+                the icon XOR bitmask.
+                */
+                if (!iconInfo.hbmColor)
+                    m_szIcon.cy /= 2;
+            }
+            DeleteObject(iconInfo.hbmMask);
+            DeleteObject(iconInfo.hbmColor);
+        }
+    }
+    EnsureWindowTheme(TRUE);
+
+    return S_OK;
+}
+
+LRESULT CTrayShowDesktopButton::OnClick(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+{
+    // The actual action can be delayed as an expected behaviour.
+    // But a too late action is an unexpected behaviour.
```

Comment:
Isn't this whole ordeal just the implementation of double-click support? If so, why not just handling that aspect in a `WM_LBUTTONDBLCLK` handler instead?

Reply:
No - this is about a single click, not a double-click.

Reply:
Then this whole delayed ordeal is weird, tbh.

Reply:
Agreed.
- On one hand...even on Windows, the button sometimes takes a weirdly long time to do its job.
- On the other hand...*why*? Is this related to that, or is this an entirely different matter?

That all being said...weirdness about what the button *does* is kind of out of scope of a PR pertaining to how the button *looks*, no?

Reply:
Ah yes ok I _may_ understand now why this is done.
Showing the desktop is done by asking all the windows to minimize (and then, restoring the desktop is done by asking all the windows to restore as they used to be). This operation takes time, and if the user plays by clicking many times per second the button, or is "a bit" slow on the contrary, things could go unexpected (for the user).

Reply:
Idk...but again, is that *really* within the scope of *this* PR?

Reply:
I was just trying to think about a rational explanation of this weird handling.

Reply:
I agree that it would be good to address/document in a better way...but it's not worse *now* than it was *before* my changes, so...are you planning to hold up the PR over this comment, or can I mark this as resolved?